ci: workflow improvements

Mirko-von-Leipzig opened the pull request:
Essentially a copy of [miden-node/#426](https://github.com/0xPolygonMiden/miden-node/pull/426).

- Limit comments to 100 character limit
- Cache ci artifacts
- Limit ci concurrency to the latest commit per PR

Closes #568

bobbinth commented on rustfmt.toml:
Let's also add
```toml
format_code_in_doc_comments = true
format_macro_matchers = true
hex_literal_case = "Lower"
```
This will make it more consistent with the latest version of this file in `miden-base`.

bobbinth commented on .github/workflows/build.yml:
```diff
@@ -18,6 +23,9 @@ jobs:
         toolchain: [stable, nightly]
     steps:
       - uses: actions/checkout@main
+      - uses: Swatinem/rust-cache@v2
+        with:
+          save-if: ${{ github.event_name == 'push' && github.ref == 'refs/heads/next' }}
```
Could be good to add a quick comment explaining what this does.

Mirko-von-Leipzig replied:
This actually made me go look into the cache effectiveness. 

I think caching is quite ineffective for many of our jobs because we used nightly for almost all of them. And caching doesn't work on a moving target.

I don't think there is much advantage to using nightly for these jobs, except `rustfmt` as it specifically has extra options available. `clippy` and `check` should probably be using stable?

bobbinth replied:
So, the solution is then to switch everything except for `rustfmt` tasks to `stable` - right?

Mirko-von-Leipzig replied:
Hmm. Okay it is a bit more awkward than that. `make lint` runs `fmt`, `clippy` and `cargo fix`. Which means it now combines stable and nightly builds. But that's probably okay since fmt is "cheap".

bobbinth commented on .github/workflows/build.yml:
```diff
@@ -32,6 +40,9 @@ jobs:
         toolchain: [stable, nightly]
     steps:
       - uses: actions/checkout@main
+      - uses: Swatinem/rust-cache@v2
+        with:
+          save-if: ${{ github.event_name == 'push' && github.ref == 'refs/heads/next' }}
```
Could be good to add a quick comment explaining what this does.

bobbinth reviewed (approved):
Looks good! Thank you! I left a couple of minor nits inline.

Mirko-von-Leipzig commented:
I've rebased and also
- removed `nightly` from `clippy` and `fix` makefile jobs
- simplified/removed matrix ci job setups where it made sense (to me)

bobbinth commented on Makefile:
```diff
@@ -48,6 +48,10 @@ doc-serve: ## Serves documentation site
 
 # --- testing -------------------------------------------------------------------------------------
 
+.PHONY: test-build
+test-build: ## Build the test binary
+	$(DEBUG_ASSERTIONS) cargo nextest run --cargo-profile test-release --features concurrent,testing --filter-expr "not test(prove)" --no-run
```
If we use `--no-run` here, why do we also need `filter-expr` to filter out some tests?

Mirko-von-Leipzig replied:
We don't; I'm just bad at copy-pasting :) Thanks!

bobbinth commented on .github/workflows/build.yml:
```diff
@@ -12,28 +17,30 @@ jobs:
   async:
     name: Build using `async` feature
```
minor nit: since all our job names start with lower case, we should probably change this to "build using..." and similar comment for "build for non-std".

bobbinth reviewed (approved):
Looks good! Thank you! I left just a couple of minor comments inline.

I think we should probably wait until https://github.com/0xPolygonMiden/miden-base/pull/826 is merged and then rebase/merge this one (my assumption is that rebasing this one should be easier than doing it the other way around).

Mirko-von-Leipzig commented:
> Looks good! Thank you! I left just a couple of minor comments inline.
> 
> I think we should probably wait until #826 is merged and then rebase/merge this one (my assumption is that rebasing this one should be easier than doing it the other way around).

Thanks, yes I'll wait then. The only conflict here will be due to the rustfmt commit, which one can simply drop and re-apply separately.
